src/librbd/cache/SharedReadOnlyObjectDispatch.cc

```diff
@@ -0,0 +1,168 @@
+    const std::string &oid, uint64_t object_no, uint64_t object_off,
+    uint64_t object_len, librados::snap_t snap_id, int op_flags,
+    const ZTracer::Trace &parent_trace, ceph::bufferlist* read_data,
+    io::ExtentMap* extent_map, int* object_dispatch_flags,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << object_len << dendl;
+
+  // if any failse, reads will go to rados
+  if(!m_cache_client->is_session_work() || m_cache_client == nullptr ||
+     m_object_store == nullptr || !m_initialzed) {
+    ldout(cct, 5) << "SRO cache client session failed " << dendl;
+    return false;
+  }
+
+  CacheGenContextURef ctx = make_gen_lambda_context<ObjectCacheRequest*,
+                                     std::function<void(ObjectCacheRequest*)>>
+   ([this, snap_id, read_data, dispatch_result, on_dispatched,
+      oid, object_off, object_len](ObjectCacheRequest* ack) {
+     handle_read_cache(ack, object_off, object_len, read_data,
+                       dispatch_result, on_dispatched);
+  });
+
+  m_cache_client->lookup_object(m_image_ctx->data_ctx.get_namespace(),
+                                m_image_ctx->data_ctx.get_id(),
+                                (uint64_t)snap_id, oid, std::move(ctx));
+  return true;
+}
+
+template <typename I, typename C>
+void SharedReadOnlyObjectDispatch<I, C>::handle_read_cache(
+    ObjectCacheRequest* ack, uint64_t read_off,
+    uint64_t read_len, ceph::bufferlist* read_data,
+    io::DispatchResult* dispatch_result, Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  if(ack->type != RBDSC_READ_REPLY) {
+    // go back to read rados
+    *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+    on_dispatched->complete(0);
+    return;
+  }
+
+  ceph_assert(ack->type == RBDSC_READ_REPLY);
+  std::string file_path = ((ObjectCacheReadReplyData*)ack)->cache_path;
+  ceph_assert(file_path != "");
+
+  // try to read from parent image cache
+  int r = m_object_store->read_object(file_path, read_data, read_off, read_len, on_dispatched);
+  if(r < 0) {
+    // cache read error, fall back to read rados
+    *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
+    on_dispatched->complete(0);
+  }
+
+  *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
+  //TODO(): complete in syncfile
+  on_dispatched->complete(r);
+}
+
+template <typename I, typename C>
+int SharedReadOnlyObjectDispatch<I, C>::handle_register_client(bool reg) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  if (reg) {
+    ldout(cct, 20) << "SRO cache client open cache handler" << dendl;
+    m_object_store = new SharedPersistentObjectCacher<I>(m_image_ctx, m_image_ctx->shared_cache_path);
+  }
+  return 0;
+}
+
+template <typename I, typename C>
+void SharedReadOnlyObjectDispatch<I, C>::client_handle_request(std::string msg) {
```

Comment:
Nit: is this used?

Reply:
thanks for jason. i have cleaned up all useless codes.